Added ItemsAdder Block Support

Teenkung123 opened the pull request:
In this request, I added support for ItemsAdder block generation by specifying the ItemsAdder custom block name, space, ID, and probability like this:

![image](https://user-images.githubusercontent.com/51858378/235899209-73c4a095-d49f-46d3-83b4-e60d1bd2d98f.png)

(It's my first time opening a pull request, so if there's anything I need to fix, please let me know.)

HSGamer commented:
Check out the `OneBlockCustomBlockCreator` and #283 

I've added a creator so that you can just make an addon to register your `OneBlockCustomBlock`, with the catch that the format for `blocks` is different.

Teenkung123 commented:
I will try doing that, but I'm not sure if it will work because I have to place ItemsAdder's custom block using customBlock.place(); otherwise, it won't recognize it and will think it's just a normal noteblock with the wrong note.

https://itemsadder.devs.beer/developers/java-api/examples

```java
CustomBlock customBlock = CustomBlock.getInstance("ruby_ore");
if(customBlock != null) //not needed if you're sure the blocks exists.
{
    customBlock.place(location);
}
else
{
    // Custom block not found in ItemsAdder configurations!
}
```

HSGamer commented:
You can use `block.getLocation()` to get the location and use that method.
Just make sure to check the custom block when registering on the creator (return empty if it's not found)

Teenkung123 commented:
I made it check on load whether the custom blocks exist in ItemsAdder or not. If they don't exist, it will log an error to the console saying the custom block does not exist. If they do exist, it will add the NamespaceID to OneBlockObject.

```java
            for (String material : blocks.getKeys(false)) {
                if (Material.getMaterial(material) != null) {
                    addMaterial(obPhase, material, Objects.toString(blocks.get(material)));
                } else {
                    if (addon.hasItemsAdder()) {
                        CustomBlock block = CustomBlock.getInstance(material);
                        if (block != null) {
                            addItemsAdderBlock(obPhase, material, Objects.toString(blocks.get(material)));
                        } else {
                            addon.logError("Bad block material in " + obPhase.getPhaseName() + ": " + material);
                        }
                    } else {
                        addon.logError("Bad block material in " + obPhase.getPhaseName() + ": " + material);
                    }
                }
            }
```

In the OneBlockObject class, I added a String variable that will store the NamespaceID of the CustomBlock. 
In the BlockListener, I added a secondary check to ensure that the blocks really exist before placing the CustomBlock.

```java
        if (nextBlock.isCustomBlock()) {
            nextBlock.getCustomBlock().setBlock(block);
            return;
        } else if (nextBlock.isItemsAdderBlock()) {
            //Get Custom Block from ItemsAdder and place it
            CustomBlock cBlock = CustomBlock.getInstance(nextBlock.getItemsAdderBlock());
            if (cBlock != null) {
                cBlock.place(block.getLocation());
                return;
            }
        }
```

tastybento commented on src/main/java/world/bentobox/aoneblock/AOneBlock.java:
```diff
@@ -34,22 +35,28 @@
     private final Config<Settings> configObject = new Config<>(this, Settings.class);
-    private BlockListener blockListener;
-    private OneBlocksManager oneBlockManager;
+    private static BlockListener blockListener;
```
Why did this become static?

tastybento commented on src/main/java/world/bentobox/aoneblock/AOneBlock.java:
```diff
@@ -34,22 +35,28 @@
-    private BlockListener blockListener;
-    private OneBlocksManager oneBlockManager;
+    private static BlockListener blockListener;
+    private static OneBlocksManager oneBlockManager;
```
Why did this become static?

tastybento commented on src/main/java/world/bentobox/aoneblock/AOneBlock.java:
```diff
@@ -79,18 +86,8 @@ private boolean loadSettings() {
-            return;
-        }
-        registerListener(blockListener);
+        loadData(true);
```
This parameter is hardcoded true. Why is it needed?

tastybento commented on src/main/java/world/bentobox/aoneblock/AOneBlock.java:
```diff
@@ -106,6 +103,24 @@ public void onEnable() {
     }
 
+    //Load some of Manager
+    public void loadData(boolean register) {
```
I see this is called by src/main/java/world/bentobox/aoneblock/listeners/ItemsAdderListener.java but not sure why it needs the parameter.

tastybento commented on src/main/java/world/bentobox/aoneblock/listeners/ItemsAdderListener.java:
```diff
@@ -0,0 +1,20 @@
+import org.bukkit.event.Listener;
+import world.bentobox.aoneblock.AOneBlock;
+
+public class ItemsAdderListener implements Listener {
```
Add some JavaDoc to explain what this class is for please.

tastybento reviewed (changes requested):
Please see the review comments. Thanks for the PR.

Teenkung123 commented:
Thanks for the review. I have resolved all of the issues in the [following commit](https://github.com/BentoBoxWorld/AOneBlock/pull/317/commits/d63a744d85fef1150798f2dd9a6ccde00d5d0922).
